pallets/foreign-investments/src/lib.rs

```diff
@@ -422,35 +380,673 @@ pub mod pallet {
+				let info = maybe_info.as_mut().ok_or(Error::<T>::InfoNotFound)?;
+
+				info.swapped_amount.ensure_add_assign(swapped_amount)?;
+				if pending_amount.is_zero() {
+					let redemption = T::Investment::redemption(&who, investment_id)?;
+
+					T::CollectedForeignRedemptionHook::notify_status_change(
+						(who.clone(), investment_id),
+						ExecutedForeignCollect {
+							currency: info.base.foreign_currency,
+							amount_currency_payout: info.swapped_amount,
+							amount_tranche_tokens_payout: info.collected_tranche_tokens(),
+							amount_remaining: redemption,
+						},
+					)?;
+
+					info.pending_tranche_tokens
+						.ensure_sub_assign(info.collected_tranche_tokens())?;
+
+					info.base.collected = CollectedAmount::default();
+					info.swapped_amount = T::Balance::default();
+
+					if info.pending_tranche_tokens.is_zero() {
+						*maybe_info = None;
+					}
+				}
+
+				Ok(())
+			})
+		}
+	}
+
+	impl<T: Config> ForeignInvestment<T::AccountId> for Pallet<T> {
+		type Amount = T::Balance;
+		type CurrencyId = T::CurrencyId;
+		type Error = DispatchError;
+		type InvestmentId = T::InvestmentId;
+
+		fn increase_foreign_investment(
+			who: &T::AccountId,
+			investment_id: T::InvestmentId,
+			foreign_amount: T::Balance,
+			foreign_currency: T::CurrencyId,
+		) -> DispatchResult {
+			// NOTE: This line will be removed with market ratios
+			let pool_amount = T::CurrencyConverter::stable_to_stable(
+				Self::pool_currency_of(investment_id)?,
+				foreign_currency,
+				foreign_amount,
+			)?;
+
+			ForeignInvestmentInfo::<T>::mutate(&who, investment_id, |info| {
+				let info = info.get_or_insert(InvestmentInfo::new(foreign_currency)?);
+
+				info.base.ensure_same_foreign(foreign_currency)?;
+				info.total_pool_amount.ensure_add_assign(pool_amount)?;
+
+				Ok::<_, DispatchError>(())
+			})?;
+
+			Self::apply_swap_and_notify(
+				who,
+				investment_id,
+				Action::Investment,
+				Swap {
+					currency_in: Self::pool_currency_of(investment_id)?,
+					currency_out: foreign_currency,
+					amount_in: pool_amount,
+				},
+			)
+		}
+
+		fn decrease_foreign_investment(
+			who: &T::AccountId,
+			investment_id: T::InvestmentId,
+			foreign_amount: T::Balance,
+			foreign_currency: T::CurrencyId,
+		) -> DispatchResult {
+			// NOTE: This line will be removed with market ratios
+			let pool_amount = T::CurrencyConverter::stable_to_stable(
+				Self::pool_currency_of(investment_id)?,
+				foreign_currency,
+				foreign_amount,
+			)?;
+
+			ForeignInvestmentInfo::<T>::mutate(&who, investment_id, |info| {
+				let info = info.as_mut().ok_or(Error::<T>::InfoNotFound)?;
+
+				info.base.ensure_same_foreign(foreign_currency)?;
+				info.total_pool_amount.ensure_sub_assign(pool_amount)?;
+
+				Ok::<_, DispatchError>(())
+			})?;
+
+			T::Investment::update_investment(
+				who,
+				investment_id,
+				T::Investment::investment(who, investment_id)?.ensure_sub(pool_amount)?,
+			)?;
```

Comment:
That looks odd. Why would you direclty call `update_investment`? Where does the knowledge come from what has already been swapped and fulfilled? 

I the above `increase_` logic, the `total_pool_amount` is just increased, and here it is just decreased.

Reply:
Shouldn't this logic come after the `Self::apply_swap_and_notify` part and use the returned `status`?

Reply:
I think this will not work, as this assumes at all times that `pool_amount` before mutate has already been locked in the investments pallet. 

## Example
* Investor increases by `100 FRAX`, assume 1:1 conversion for simplicity
* FI creates swap for `100 FRAX > USDC`
* Order gets fulfilled with `50 USDC`
* `FI` updates the order of the investor to `50 USDC` in the `Investments`
* Investors decreases by `100 FRAX`

Now here is the problem IMO. This code will now try to reduce the order by `100 USDC` while it can at most decrease by `50 USDC`. Furthermore, if we are using market rates, then this becomes even harder, as even if all has been swapped, the rate is "better" now, that you want to decrease the order by too much.

Reply:
You're right. It should be done after `apply_swap_and_notify()` reading the status.

I understood this in a way that `decrease_foreign_investment()` only decreases for amounts already swapped. In such a case, the current investment already has the amount invested and swapped. But it's true that we're skipping the cases where I want to decrease a previous increase that was not swapped.

Reply:
Increasing and decreasing a foreign investment is asynchronous:

1. User signals increase
2. Apply swap
3. ASYNC: When the swap order (partially) fulfils, the swapped ...
  a. If increased swap: `pool_amount` can be invested
  b. If decreased swap: `foreign_amount` can be transferred back

Reply:
But we need to reduce the investment before swapping if we are decrementing. So, in case of decrement, we can not apply swap directly.

Reply:
> But we need to reduce the investment before swapping if we are decrementing. So, in case of decrement, we can not apply swap directly.

That depends on the current pending swap:
* If the swap is `pool -> foreign` with `swap_amount >= decrease_amount`, we don't need to decrease the investment
* If the swap is `pool -> foreign` with `swap_amount < decrease_amount`, we need to reduce the investment by `decrease_amount - swap_amount`
* Else we need to decrease the investment by `decrease_amount`

Reply:
Here is the solution: [eca018c](https://github.com/centrifuge/centrifuge-chain/pull/1698/commits/eca018cb1db5bd038e5ac9ff3e8189d678a7bb3d)

Reply:
That LGTM!